lib/puppet/environments.rb

```diff
@@ -230,20 +233,22 @@ def create_environment(name)
       env
     end
 
-    def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
+      return envdir
     end
 
-    def valid_directories
+    def valid_environments
+      envs = []
       if Puppet::FileSystem.directory?(@environment_dir)
-        Puppet::FileSystem.children(@environment_dir).select do |child|
-          valid_directory?(child)
+        Puppet::FileSystem.children(@environment_dir).each do |child|
```

Comment:
why go from `select` to appending to an array here?

Reply:
another remnant of my in-between refactor. but I guess this now returns just env names, rather than whole directories?

Reply:
I think if you change it back to `select` and `env_name if validated_directory(child)` will return nil if it doesn't match and select won't include it in the output (and if everything is removed it will return an empty array. Removing the need for `envs`.

Reply:
yeah, I can go back to `select`

Reply:
hmm actually I don't know if that will work because I'm still not trying to return the actual `child`s.. might need to `map`?

Reply:
oh good point

Reply:
updated